File: prob10/prob10.java
import java.util.Arrays;

public class prob10 {

    public static final int MAX = 2000000;

    /** Sieves to find all primes up to n.

	@param maximum number
	@return a boolean[], where a[i] tells if i is prime 
    */
    public static boolean[] sieve(int n) 
    {
	boolean[] primes = new boolean[n];

	Arrays.fill(primes,true);
	primes[0]=false;
	primes[1]=false;

	// iterate thru every number.
	for(int i=2; i < n; ++i) {
	    // if we're encountering a new prime ...
	    if(primes[i]==true) {
		// .. mark its multiples as non-prime
		for(int j=(i+i); j < n; j=j+i) {
		    primes[j]=false;
		}
	    }
	}
	return primes;
    }
    
    public static void main(String[] args) {
	boolean[] primes = sieve(MAX);

	long sum = 0;
	for(int i=1; i < MAX; ++i) {
	    if(primes[i] == true)
		sum += i;
	}
	System.out.println("The primes up to " + MAX + " sum to " + sum);

	int count=1;
	int i =1;
	while(count < 1000) {
	    if(primes[i] ==true)
		{
		    System.out.println("The "+ count++ +"th prime is " + i);
		}
	    i++;
	}
    }
}

File: prob10/prob10.cpp
#include<iostream>
#include<vector>
#include<list>

#include <boost/cstdint.hpp>
using boost::uint64_t;

using std::list;
using std::vector;
using std::cout;
using std::endl;


/* prints every item in C */
template<typename C> 
void print(C const & c) 
{ 
  std::cout << "{";
  for(typename C::const_iterator i=c.begin(),e=c.end(); i!=e; ++i) { 
    cout << *i << ", "; 
  } 
  std::cout << "}";
} 

/* returns a vector, with primes[i] telling if i is prime for i < n 

   Uses the sieve of eratosthenes.
*/
std::vector<bool> primeFlags(const long n) 
{
  vector<bool> primes(n,true); 
  primes[1]=false;

  // iterate thru every number
  for(long i=2; i < n; ++i) {
    // if we're encountering a new prime
    if(primes[i] == true) {
      for(long j=(i+i); j < n; j=j+i) {
	primes[j]=false;
      }
    }
  }
  return primes;
}

/* Returns a list of primes < n */
list<int> primesTo(const unsigned int n) {
  list<int> primes;
  
  vector<bool> v = primeFlags(n);
  for(unsigned int i=2; i < n; ++i) {
    if(v[i] == true) {
      primes.push_back(i);
    }
  }
  return primes;
}

/* Sums primes under n */
uint64_t sumPrimesTo(const unsigned long n) {
  uint64_t sum=0;

  const vector<bool> primes = primeFlags(n);
  for(unsigned long i = 2; i < n; ++i) {
    if(primes[i] == true) {
      sum += i;
    }
  }
  
  return sum;
}


int main()
{
  const unsigned long limit = 2000000;

  cout << "sum of primes < " << limit << " = " << sumPrimesTo(limit) << endl;
  cout << "primeFlags(20) = ";
  //  vector<bool> primes = sieve(21);
  //  print(primes);
  print(primeFlags(21));
  cout << endl;

  cout << "primesTo(20) = ";
  print(primesTo(20));
  cout << endl;

}
